File: dotnet/Rezel/Grammars/RustGrammar.cs
using Rezel.Common;
using Rezel.Lr;
using Rezel.Highlight;

namespace Rezel.Grammars;

public static class RustGrammar
{
    private const int B = 98;
    private const int E = 101;
    private const int F = 102;
    private const int R = 114;
    private const int BigE = 69;
    private const int Zero = 48;
    private const int Dot = 46;
    private const int Plus = 43;
    private const int Minus = 45;
    private const int Hash = 35;
    private const int Quote = 34;
    private const int Pipe = 124;
    private const int LessThan = 60;
    private const int GreaterThan = 62;

    private static bool IsNum(int ch) => ch >= 48 && ch <= 57;
    private static bool IsNum_(int ch) => IsNum(ch) || ch == 95;
    private static bool IsWord(int ch) => (ch >= 65 && ch <= 90) || (ch >= 97 && ch <= 122) || ch == 95 || ch >= 128;

    private static ITokenizer MakeRustExternalTokenizer(string name, Dictionary<string, int> terms)
    {
        if (name == "literalTokens")
        {
            var Float = terms["Float"];
            var RawString = terms["RawString"];

            return new ExternalTokenizer((input, _) =>
            {
                if (IsNum(input.Next))
                {
                    var isFloat = false;
                    do { input.Advance(); } while (IsNum_(input.Next));
                    if (input.Next == Dot)
                    {
                        isFloat = true;
                        input.Advance();
                        if (IsNum(input.Next))
                        {
                            do { input.Advance(); } while (IsNum_(input.Next));
                        }
                        else if (input.Next == Dot || input.Next > 0x7f || IsWord(input.Next))
                        {
                            return;
                        }
                    }
                    if (input.Next == E || input.Next == BigE)
                    {
                        isFloat = true;
                        input.Advance();
                        if (input.Next == Plus || input.Next == Minus) input.Advance();
                        if (!IsNum_(input.Next)) return;
                        do { input.Advance(); } while (IsNum_(input.Next));
                    }
                    if (input.Next == F)
                    {
                        var after = input.Peek(1);
                        if ((after == Zero + 3 && input.Peek(2) == Zero + 2) ||
                            (after == Zero + 6 && input.Peek(2) == Zero + 4))
                        {
                            input.Advance();
                            input.Advance();
                            input.Advance();
                            isFloat = true;
                        }
                        else
                        {
                            return;
                        }
                    }
                    if (isFloat) input.AcceptToken(Float);
                }
                else if (input.Next == B || input.Next == R)
                {
                    if (input.Next == B) input.Advance();
                    if (input.Next != R) return;
                    input.Advance();
                    var count = 0;
                    while (input.Next == Hash) { count++; input.Advance(); }
                    if (input.Next != Quote) return;
                    input.Advance();
                    while (true)
                    {
                        if (input.Next < 0) return;
                        var isQuote = input.Next == Quote;
                        input.Advance();
                        if (isQuote)
                        {
                            var matched = true;
                            for (var i = 0; i < count; i++)
                            {
                                if (input.Next != Hash) { matched = false; break; }
                                input.Advance();
                            }
                            if (matched)
                            {
                                input.AcceptToken(RawString);
                                return;
                            }
                        }
                    }
                }
            });
        }

        if (name == "closureParam")
        {
            var closureParamDelim = terms["closureParamDelim"];
            return new ExternalTokenizer((input, _) =>
            {
                if (input.Next == Pipe) input.AcceptToken(closureParamDelim, endOffset: 1);
            });
        }

        if (name == "tpDelim")
        {
            var tpOpen = terms["tpOpen"];
            var tpClose = terms["tpClose"];
            return new ExternalTokenizer((input, _) =>
            {
                if (input.Next == LessThan) input.AcceptToken(tpOpen, endOffset: 1);
                else if (input.Next == GreaterThan) input.AcceptToken(tpClose, endOffset: 1);
            });
        }

        throw new ArgumentException($"Unknown Rust external tokenizer: {name}");
    }

    private static readonly NodePropSource RustHighlighting = HighlightUtil.StyleTags(new Dictionary<string, object>
    {
        ["const macro_rules struct union enum type fn impl trait let static"] = Tags.DefinitionKeyword,
        ["mod use crate"] = Tags.ModuleKeyword,
        ["pub unsafe async mut extern default move"] = Tags.ModifierTag,
        ["for if else loop while match continue break return await"] = Tags.ControlKeyword,
        ["as in ref"] = Tags.OperatorKeyword,
        ["where _ crate super dyn"] = Tags.Keyword,
        ["self"] = Tags.Self,
        ["String"] = Tags.String,
        ["Char"] = Tags.Character,
        ["RawString"] = Tags.Special(Tags.String),
        ["Boolean"] = Tags.Bool,
        ["Identifier"] = Tags.VariableName,
        ["CallExpression/Identifier"] = Tags.Function(Tags.VariableName),
        ["BoundIdentifier"] = Tags.Definition(Tags.VariableName),
        ["FunctionItem/BoundIdentifier"] = Tags.Function(Tags.Definition(Tags.VariableName)),
        ["LoopLabel"] = Tags.LabelName,
        ["FieldIdentifier"] = Tags.PropertyName,
        ["CallExpression/FieldExpression/FieldIdentifier"] = Tags.Function(Tags.PropertyName),
        ["Lifetime"] = Tags.Special(Tags.VariableName),
        ["ScopeIdentifier"] = Tags.Namespace,
        ["TypeIdentifier"] = Tags.TypeName,
        ["MacroInvocation/Identifier MacroInvocation/ScopedIdentifier/Identifier"] = Tags.MacroName,
        ["MacroInvocation/TypeIdentifier MacroInvocation/ScopedIdentifier/TypeIdentifier"] = Tags.MacroName,
        ["\"!\""] = Tags.MacroName,
        ["UpdateOp"] = Tags.UpdateOperator,
        ["LineComment"] = Tags.LineComment,
        ["BlockComment"] = Tags.BlockComment,
        ["Integer"] = Tags.Integer,
        ["Float"] = Tags.Float,
        ["ArithOp"] = Tags.ArithmeticOperator,
        ["LogicOp"] = Tags.LogicOperator,
        ["BitOp"] = Tags.BitwiseOperator,
        ["CompareOp"] = Tags.CompareOperator,
        ["="] = Tags.DefinitionOperator,
        [".. ... => ->"] = Tags.Punctuation,
        ["( )"] = Tags.Paren,
        ["[ ]"] = Tags.SquareBracket,
        ["{ }"] = Tags.Brace,
        [". DerefOp"] = Tags.DerefOperator,
        ["&"] = Tags.Operator,
        [", ; ::"] = Tags.Separator,
        ["Attribute/..."] = Tags.Meta,
    });

    private static LRParser? _parser;
    public static LRParser Parser => _parser ??= CreateParser();

    private static LRParser CreateParser()
    {
        var termTable = RustParserData.TermTable!;
        var externals = new Dictionary<string, ITokenizer>
        {
            ["closureParam"] = MakeRustExternalTokenizer("closureParam", termTable),
            ["tpDelim"] = MakeRustExternalTokenizer("tpDelim", termTable),
            ["literalTokens"] = MakeRustExternalTokenizer("literalTokens", termTable),
        };
        var spec = RustParserData.MakeSpec(
            externals: externals,
            propSources: [RustHighlighting]
        );
        return new LRParser(spec);
    }
}

File: swift/Sources/Rezel/Grammars/RustGrammar.swift
import Foundation

private let _b = 98
private let _e = 101
private let _f = 102
private let _r = 114
private let _E = 69
private let Zero = 48
private let Dot = 46
private let Plus = 43
private let Minus = 45
private let Hash = 35
private let Quote = 34
private let Pipe = 124
private let LessThan = 60
private let GreaterThan = 62

private func isNum(_ ch: Int) -> Bool {
	return ch >= 48 && ch <= 57
}

private func isNum_(_ ch: Int) -> Bool {
	return isNum(ch) || ch == 95
}

private func isWordChar(_ ch: Int) -> Bool {
	guard let scalar = UnicodeScalar(ch) else { return false }
	return scalar.properties.isAlphabetic || scalar.properties.numericType != nil || ch == 95
}

private func makeRustExternalTokenizer(name: String, terms: [String: Int]) -> TokenizerProtocol {
	if name == "literalTokens" {
		let Float = terms["Float"]!
		let RawString = terms["RawString"]!
		return ExternalTokenizer { input, _ in
			if isNum(input.next) {
				var isFloat = false
				repeat {
					input.advance()
				} while isNum_(input.next)
				if input.next == Dot {
					isFloat = true
					input.advance()
					if isNum(input.next) {
						repeat {
							input.advance()
						} while isNum_(input.next)
					} else if input.next == Dot || input.next > 0x7F || isWordChar(input.next) {
						return
					}
				}
				if input.next == _e || input.next == _E {
					isFloat = true
					input.advance()
					if input.next == Plus || input.next == Minus { input.advance() }
					if !isNum_(input.next) { return }
					repeat {
						input.advance()
					} while isNum_(input.next)
				}
				if input.next == _f {
					let after = input.peek(1)
					if (after == Zero + 3 && input.peek(2) == Zero + 2) ||
						(after == Zero + 6 && input.peek(2) == Zero + 4)
					{
						input.advance(3)
						isFloat = true
					} else {
						return
					}
				}
				if isFloat { input.acceptToken(Float) }
			} else if input.next == _b || input.next == _r {
				if input.next == _b { input.advance() }
				if input.next != _r { return }
				input.advance()
				var count = 0
				while input.next == Hash {
					count += 1; input.advance()
				}
				if input.next != Quote { return }
				input.advance()
				contentLoop: while true {
					if input.next < 0 { return }
					let isQuote = input.next == Quote
					input.advance()
					if isQuote {
						for _ in 0 ..< count {
							if input.next != Hash { continue contentLoop }
							input.advance()
						}
						input.acceptToken(RawString)
						return
					}
				}
			}
		}
	}

	if name == "closureParam" {
		let closureParamDelim = terms["closureParamDelim"]!
		return ExternalTokenizer { input, _ in
			if input.next == Pipe { input.acceptToken(closureParamDelim, endOffset: 1) }
		}
	}

	if name == "tpDelim" {
		let tpOpen = terms["tpOpen"]!
		let tpClose = terms["tpClose"]!
		return ExternalTokenizer { input, _ in
			if input.next == LessThan {
				input.acceptToken(tpOpen, endOffset: 1)
			} else if input.next == GreaterThan {
				input.acceptToken(tpClose, endOffset: 1)
			}
		}
	}

	fatalError("Unknown Rust external tokenizer: \(name)")
}

private nonisolated(unsafe) let definition = hlTags["definition"] as! (Tag) -> Tag
private nonisolated(unsafe) let functionMod = hlTags["function"] as! (Tag) -> Tag
private nonisolated(unsafe) let specialMod = hlTags["special"] as! (Tag) -> Tag

nonisolated(unsafe) let rustHighlighting = styleTags([
	"const macro_rules struct union enum type fn impl trait let static": hlTags["definitionKeyword"] as Any,
	"mod use crate": hlTags["moduleKeyword"] as Any,
	"pub unsafe async mut extern default move": hlTags["modifier"] as Any,
	"for if else loop while match continue break return await": hlTags["controlKeyword"] as Any,
	"as in ref": hlTags["operatorKeyword"] as Any,
	"where _ crate super dyn": hlKeyword,
	"self": hlTags["self"] as Any,
	"String": hlString,
	"Char": hlTags["character"] as Any,
	"RawString": specialMod(hlString),
	"Boolean": hlTags["bool"] as Any,
	"Identifier": hlTags["variableName"] as Any,
	"CallExpression/Identifier": functionMod(hlTags["variableName"] as! Tag),
	"BoundIdentifier": definition(hlTags["variableName"] as! Tag),
	"FunctionItem/BoundIdentifier": functionMod(definition(hlTags["variableName"] as! Tag)),
	"LoopLabel": hlTags["labelName"] as Any,
	"FieldIdentifier": hlPropertyName,
	"CallExpression/FieldExpression/FieldIdentifier": functionMod(hlPropertyName),
	"Lifetime": specialMod(hlTags["variableName"] as! Tag),
	"ScopeIdentifier": hlTags["namespace"] as Any,
	"TypeIdentifier": hlTypeName,
	"MacroInvocation/Identifier MacroInvocation/ScopedIdentifier/Identifier": hlTags["macroName"] as Any,
	"MacroInvocation/TypeIdentifier MacroInvocation/ScopedIdentifier/TypeIdentifier": hlTags["macroName"] as Any,
	"\"!\"": hlTags["macroName"] as Any,
	"UpdateOp": hlTags["updateOperator"] as Any,
	"LineComment": hlTags["lineComment"] as Any,
	"BlockComment": hlTags["blockComment"] as Any,
	"Integer": hlTags["integer"] as Any,
	"Float": hlTags["float"] as Any,
	"ArithOp": hlTags["arithmeticOperator"] as Any,
	"logicOp": hlTags["logicOperator"] as Any,
	"BitOp": hlTags["bitwiseOperator"] as Any,
	"CompareOp": hlTags["compareOperator"] as Any,
	"=": hlTags["definitionOperator"] as Any,
	".. ... => ->": hlPunctuation,
	"( )": hlTags["paren"] as Any,
	"[ ]": hlTags["squareBracket"] as Any,
	"{ }": hlTags["brace"] as Any,
	". DerefOp": hlTags["derefOperator"] as Any,
	"&": hlTags["operator"] as Any,
	", ; ::": hlTags["separator"] as Any,
	"Attribute/...": hlMeta,
])

public let rustParser: LRParser = {
	let externals: [String: TokenizerProtocol] = [
		"literalTokens": makeRustExternalTokenizer(name: "literalTokens", terms: RustParserData.termTable),
		"closureParam": makeRustExternalTokenizer(name: "closureParam", terms: RustParserData.termTable),
		"tpDelim": makeRustExternalTokenizer(name: "tpDelim", terms: RustParserData.termTable),
	]
	let spec = RustParserData.makeSpec(
		externals: externals,
		propSources: [rustHighlighting]
	)
	return LRParser(spec: spec)
}()
